File: golang/digest.go
package httpsig

import (
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"fmt"
	"hash"
	"strings"
)

// DigestAlgorithm identifies a digest algorithm for Content-Digest (RFC 9530).
type DigestAlgorithm string

const (
	DigestSHA256 DigestAlgorithm = "sha-256"
	DigestSHA512 DigestAlgorithm = "sha-512"
)

// ContentDigest computes a Content-Digest header value per RFC 9530.
func ContentDigest(body []byte, alg DigestAlgorithm) (string, error) {
	var h hash.Hash
	switch alg {
	case DigestSHA256:
		h = sha256.New()
	case DigestSHA512:
		h = sha512.New()
	default:
		return "", fmt.Errorf("unsupported digest algorithm: %s", alg)
	}
	h.Write(body)
	digest := h.Sum(nil)
	return fmt.Sprintf("%s=:%s:", alg, base64.StdEncoding.EncodeToString(digest)), nil
}

// VerifyContentDigest checks that a Content-Digest header value matches the body.
func VerifyContentDigest(body []byte, headerValue string) (bool, error) {
	// Parse all digest values from the header
	// Format: sha-256=:base64:, sha-512=:base64:
	parts := strings.Split(headerValue, ",")
	for _, part := range parts {
		part = strings.TrimSpace(part)
		eqIdx := strings.Index(part, "=:")
		if eqIdx < 0 {
			continue
		}
		algStr := strings.TrimSpace(part[:eqIdx])
		rest := part[eqIdx+2:]
		colonIdx := strings.LastIndex(rest, ":")
		if colonIdx < 0 {
			continue
		}
		b64 := rest[:colonIdx]
		expected, err := base64.StdEncoding.DecodeString(b64)
		if err != nil {
			continue
		}

		alg := DigestAlgorithm(algStr)
		var h hash.Hash
		switch alg {
		case DigestSHA256:
			h = sha256.New()
		case DigestSHA512:
			h = sha512.New()
		default:
			continue
		}
		h.Write(body)
		actual := h.Sum(nil)

		if len(actual) != len(expected) {
			return false, nil
		}
		match := true
		for i := range actual {
			if actual[i] != expected[i] {
				match = false
				break
			}
		}
		if match {
			return true, nil
		}
	}
	return false, nil
}

File: typescript/src/digest.ts
import * as crypto from "node:crypto";
import type { DigestAlgorithm } from "./types.js";

/** Compute a Content-Digest header value per RFC 9530. */
export function contentDigest(
  body: Uint8Array,
  alg: DigestAlgorithm,
): string {
  let hashAlg: string;
  switch (alg) {
    case "sha-256":
      hashAlg = "sha256";
      break;
    case "sha-512":
      hashAlg = "sha512";
      break;
    default:
      throw new Error(`unsupported digest algorithm: ${alg}`);
  }
  const hash = crypto.createHash(hashAlg);
  hash.update(body);
  const digest = hash.digest("base64");
  return `${alg}=:${digest}:`;
}

/** Verify a Content-Digest header value matches the body. */
export function verifyContentDigest(
  body: Uint8Array,
  headerValue: string,
): boolean {
  const parts = headerValue.split(",");
  for (const part of parts) {
    const trimmed = part.trim();
    const eqIdx = trimmed.indexOf("=:");
    if (eqIdx < 0) continue;

    const algStr = trimmed.slice(0, eqIdx).trim();
    const rest = trimmed.slice(eqIdx + 2);
    const colonIdx = rest.lastIndexOf(":");
    if (colonIdx < 0) continue;

    const b64 = rest.slice(0, colonIdx);
    let expected: Buffer;
    try {
      expected = Buffer.from(b64, "base64");
    } catch {
      continue;
    }

    let hashAlg: string;
    switch (algStr as DigestAlgorithm) {
      case "sha-256":
        hashAlg = "sha256";
        break;
      case "sha-512":
        hashAlg = "sha512";
        break;
      default:
        continue;
    }

    const hash = crypto.createHash(hashAlg);
    hash.update(body);
    const actual = hash.digest();

    if (actual.length === expected.length && crypto.timingSafeEqual(actual, expected)) {
      return true;
    }
  }
  return false;
}
